liblwgeom/lwgeom.c

```diff
@@ -1606,55 +1621,29 @@ lwgeom_remove_repeated_points_in_place(LWGEOM *geom, double tolerance)
 		}
 		case MULTIPOINTTYPE:
 		{
-			double tolsq = tolerance*tolerance;
-			uint32_t i, j, n = 0;
-			LWMPOINT *mpt = (LWMPOINT *)(geom);
-			LWPOINT **out;
-			LWPOINT *out_stack[out_stack_size];
-			int use_heap = (mpt->ngeoms > out_stack_size);
-
-			/* No-op on empty */
-			if (mpt->ngeoms < 2)
-				return geometry_modified;
-
-			/* We cannot write directly back to the multipoint */
-			/* geoms array because we're reading out of it still */
-			/* so we use a side array */
-			if (use_heap)
-				out = lwalloc(sizeof(LWMPOINT *) * mpt->ngeoms);
-			else
-				out = out_stack;
+			double tolsq = tolerance * tolerance;
+			LWMPOINT *mpt = (LWMPOINT *)geom;
+			qsort(mpt->geoms, mpt->ngeoms, sizeof(LWPOINT *), point_cmp);
 
-			/* Inefficient O(n^2) implementation */
-			for (i = 0; i < mpt->ngeoms; i++)
+			uint32_t i = 0;
+			for (uint32_t j = 1; j < mpt->ngeoms; j++)
 			{
-				int seen = 0;
-				LWPOINT *p1 = mpt->geoms[i];
-				const POINT2D *pt1 = getPoint2d_cp(p1->point, 0);
-				for (j = 0; j < n; j++)
+				POINT2D ptj, pti;
+				getPoint2d_p(mpt->geoms[i]->point, 0, &pti);
```

Comment:
can you get a _cp ("const pointer") here? it will point directly to memory in pointarray and skip copying